presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardCompactionManager.java

```diff
@@ -234,4 +230,87 @@ private boolean needsCompaction(ShardMetadata shard)
         }
         return false;
     }
+
+    @VisibleForTesting
+    public static class CompactionSetCreator
+    {
+        private final DataSize maxShardSize;
+        private final long maxShardRows;
+
+        public CompactionSetCreator(DataSize maxShardSize, long maxShardRows)
+        {
+            checkArgument(maxShardRows > 0, "maxShardRows must be > 0");
+
+            this.maxShardSize = requireNonNull(maxShardSize, "maxShardSize is null");
+            this.maxShardRows = maxShardRows;
+        }
+
+        public Set<OrganizationSet> createCompactionSets(Table tableInfo, Collection<ShardIndexInfo> shards)
+        {
+            Collection<Collection<ShardIndexInfo>> shardsByDaysBuckets = getShardsByDaysBuckets(tableInfo, shards);
+
+            ImmutableSet.Builder<OrganizationSet> compactionSets = ImmutableSet.builder();
+            for (Collection<ShardIndexInfo> shardInfos : shardsByDaysBuckets) {
+                compactionSets.addAll(buildCompactionSets(tableInfo, ImmutableSet.copyOf(shardInfos)));
+            }
+            return compactionSets.build();
+        }
+
+        private Set<OrganizationSet> buildCompactionSets(Table tableInfo, Set<ShardIndexInfo> shardIndexInfos)
+        {
+            long tableId = tableInfo.getTableId();
+            List<ShardIndexInfo> shards = shardIndexInfos.stream()
```

Comment:
Those shards should never come to this call. We filter them out ahead of time and since they are immutable there is no reason to check here again.